### PR TITLE
NGSI-LD howto

### PR DESCRIPTION
**Pull request by jmcanterafonseca**

### CODE REVIEW

File: specs/ngsi-ld_howto.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    namespace.
+-   Regular entity attributes have to be converted to JSON-LD nodes of type
+    `Property`.
+-   ref attributes (pointing to other entities) have to be converted to JSON-LD
+    nodes of type `Relationship`.
+-   The `timestamp` metadata item has to be mapped to the `observedAt` member of
+    a Property node.
+-   The `unitCode` metadata item has to be mapped to the `unitCode` member of a
+    Property node.
+-   The NGSI `DateTime` type has to be properly encoded as per the JSON-LD
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: For ease of understanding explicitly use NGSI v2

File: specs/ngsi-ld_howto.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    rules.
+-   The NGSI `geo:json` type has to be renamed to `GeoProperty`.
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: For ease of understanding explicitly use NGSI v2

File: specs/ngsi-ld_howto.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+The main advantage of JSON-LD is that it offers the capability of expanding JSON
+terms to URIs, so that vocabularies can define terms unambiguously.
+
+## Steps to migrate to JSON-LD
+
+First of all, each Data Model shall have a JSON-LD @context, providing an
+unambiguous definition by mapping terms to URIs. For practicality reasons, it is
+recommended to have a unique @context resource, containing all terms, subject to
+be used in every FIWARE Data Model, the same way as schema.org does. The
+following steps have to be followed in order to migrate existing NGSI
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: For ease of understanding explicitly use NGSI v2

File: specs/ngsi-ld_howto.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+be one, equivalent to the NGSI metadata abstraction. NGSI-LD Entities are
+represented using JSON-LD, a JSON-based serialization format for Linked Data.
+The main advantage of JSON-LD is that it offers the capability of expanding JSON
+terms to URIs, so that vocabularies can define terms unambiguously.
+
+## Steps to migrate to JSON-LD
+
+First of all, each Data Model shall have a JSON-LD @context, providing an
+unambiguous definition by mapping terms to URIs. For practicality reasons, it is
+recommended to have a unique @context resource, containing all terms, subject to
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: `@context` should be monospaced

File: specs/ngsi-ld_howto.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+## Introduction
+
+The OMA NGSI information model has been evolved to better support linked data
+(entity relationships), property graphs and semantics (exploiting the
+capabilities offered by JSON-LD). This work has been conducted under the ETSI
+ISG CIM initiative and has been branded as
+[NGSI-LD](https://www.etsi.org/deliver/etsi_gs/CIM/001_099/009/01.01.01_60/gs_CIM009v010101p.pdf).
+The main constructs of NGSI-LD are: Entity, Property and Relationship. NGSI-LD
+Entities (instances) can be the subject of Properties or Relationships. In terms
+of the traditional NGSI data model, Properties can be seen as the combination of
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: For ease of understanding explicitly use NGSI v2

File: specs/ngsi-ld_howto.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+an attribute and its value. Relationships allow to establish associations
+between instances using linked data. In practice, they are conveyed by means of
+an NGSI attribute, but with a special value (relationship’s object) which
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: Suggest:
> In practice, they are conveyed by means of a special NGSI v2 attribute with a special value

File: specs/ngsi-ld_howto.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+following steps have to be followed in order to migrate existing NGSI
+instantiations of the FIWARE Data Models to NGSI-LD:
+
+-   Entity ids have to be converted to URIs, preferably using the NGSI-LD URN
+    namespace.
+-   Regular entity attributes have to be converted to JSON-LD nodes of type
+    `Property`.
+-   ref attributes (pointing to other entities) have to be converted to JSON-LD
+    nodes of type `Relationship`.
+-   The `timestamp` metadata item has to be mapped to the `observedAt` member of
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: Suggest:

> The `timestamp` NGSI v2 metadata item

File: specs/ngsi-ld_howto.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+## Steps to migrate to JSON-LD
+
+First of all, each Data Model shall have a JSON-LD @context, providing an
+unambiguous definition by mapping terms to URIs. For practicality reasons, it is
+recommended to have a unique @context resource, containing all terms, subject to
+be used in every FIWARE Data Model, the same way as schema.org does. The
+following steps have to be followed in order to migrate existing NGSI
+instantiations of the FIWARE Data Models to NGSI-LD:
+
+-   Entity ids have to be converted to URIs, preferably using the NGSI-LD URN
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: Suggest;

> -   NGSI v2 entity `id` attributes have to be converted to URIs, preferably using the NGSI-LD URN

File: specs/ngsi-ld_howto.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        }
+    },
+    "refPointOfInterest": {
+        "type": "Reference",
+        "value": "PointOfInterest:RZ:MainSquare"
+    }
+}
+```
+
+```json
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: Add heading

File: specs/ngsi-ld_howto.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+```
+
+The figure below shows how air quality information, at a certain point of
+interest, can be conveyed using the FIWARE Data Models (involving the entity
+types `AirQualityObserved`, `PointOfInterest`) in NGSI-LD format. The new
+representation has been obtained by applying the conversion rules described
+before. Please note that the referred JSON-LD @context could be eventually
+published by FIWARE at the suggested URL.
+
+```json
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: Add heading

File: specs/ngsi-ld_howto.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        "object": "urn:ngsi-ld:PointOfInterest:RZ:MainSquare"
+    },
+    "@context": [
+        "https://schema.lab.fiware.org/ld/jsonldcontext.json",
+        "http://uri.etsi.org/ngsi-ld/ngsi-ld-core-context.jsonld"
+    ]
+}
+```
+
+```json
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: Add heading

File: specs/ngsi-ld_howto.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+happens to be a URI which points to another entity. They are similar to the
+“ref” attributes recommended the Data Models guidelines.
+
+Properties and Relationships can be the subject of other Properties or
+Relationships. Thus, in the NGSI-LD information model there are no attribute’s
+metadata, but just “properties of properties” or “properties of relationships”.
+It is not expected to have infinite graphs, and in practice, only one or two
+levels of Property or Relationship “chaining” will happen. Typically, there will
+be one, equivalent to the NGSI metadata abstraction. NGSI-LD Entities are
+represented using JSON-LD, a JSON-based serialization format for Linked Data.
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: First instance of JSON-LD should be a link

File: specs/ngsi-ld_howto.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        }
+    },
+    "@context": [
+        "https://schema.lab.fiware.org/ld/jsonldcontext.json",
+        "http://uri.etsi.org/ngsi-ld/ngsi-ld-core-context.jsonld"
+    ]
+}
+```
+
+The content of the JSON-LD @context could contain the mappings enumerated below.
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: suggest: 

> The content of the `@context` attribute contains JSON-LD  mappings such as those enumerated below.

File: specs/ngsi-ld_howto.md (hunk 12 above)
Comment (jmcanterafonseca) [reply specs/ngsi-ld_howto.md]: everything fixed at https://github.com/Fiware/dataModels/pull/375/commits/6a4d190ec9a57e47bda230c5fedc2801d63a2a91

Comment (jmcanterafonseca) [reply specs/ngsi-ld_howto.md]: https://github.com/Fiware/dataModels/pull/375/commits/d94be2f4ffab6861705c2cab1fe2711203a79fcf

Comment (jason-fox) [reply specs/ngsi-ld_howto.md]: LGTM

File: specs/ngsi-ld_howto.md (hunk 9 above)
Comment (jason-fox) [reply specs/ngsi-ld_howto.md]: LGTM

File: specs/ngsi-ld_howto.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+The FIWARE Community has already provided a simple script to migrate FIWARE NGSI
+entity representations to NGSI-LD, see
+https://github.com/Fiware/dataModels/blob/master/tools/normalized2LD.py
+
+## Example of migration to NGSI-LD
+
+The figure below shows how air quality information at a certain point of
+interest can be conveyed using the FIWARE Data Models (involving the entity
+types `AirQualityObserved`, `PointOfInterest`) in NGSI (JSON) format.
+
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply specs/ngsi-ld_howto.md]: LGTM

File: specs/ngsi-ld_howto.md (hunk 7 above)
Comment (jason-fox) [reply specs/ngsi-ld_howto.md]: NTC - this is explained in the DateTime bullet

File: specs/ngsi-ld_howto.md (hunk 10 above)
Comment (jason-fox) [reply specs/ngsi-ld_howto.md]: LGTM

File: specs/ngsi-ld_howto.md (hunk 11 above)
Comment (jason-fox) [reply specs/ngsi-ld_howto.md]: LGTM

File: specs/ngsi-ld_howto.md [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+        "http://uri.etsi.org/ngsi-ld/ngsi-ld-core-context.jsonld"
+    ]
+}
+```
+
+The content of the JSON-LD `@context` could contain the mappings enumerated
+below. Observe that `refPointOfInterest` is labelled as an `@id`, as it is
+actually pointing to another Entity (linked data). On the other hand, there are
+certain terms such as `location` or `unitCode` which are not included in the
+@context, as they pertain to the Core JSON-LD `@context` which is always
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: `@context` not in monospace

File: specs/ngsi-ld_howto.md (hunk 3 above)
Comment (jason-fox) [reply specs/ngsi-ld_howto.md]: Suggest:
>  migrate existing NGSI v2

File: specs/ngsi-ld_howto.md [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+The FIWARE Community has already provided a simple script to migrate FIWARE NGSI
+entity representations to NGSI-LD, see
+[https://github.com/Fiware/dataModels/blob/master/tools/normalized2LD.py]
+
+## Example of migration to NGSI-LD
+
+The figure below shows how air quality information at a certain point of
+interest can be conveyed using the FIWARE Data Models (involving the entity
+types `AirQualityObserved`, `PointOfInterest`) in NGSI v2 (JSON) format.
```
Comment (jason-fox) [inline specs/ngsi-ld_howto.md]: Suggest:

>in NGSI v2 ~(JSON)~ format.

i.e. remove JSON

**Pull request merged**